seqr/views/apis/anvil_workspace_api_tests.py

```diff
@@ -405,32 +477,144 @@ def test_create_project_from_workspace(self, mock_tempfile, mock_mv_file, mock_s
+            ['HG00735', 'NA19675', 'NA19678'], 'GRCh38', REQUEST_BODY)
+
+        # Test logged in locally
+        remove_token(self.manager_user)  # The user will look like having logged in locally after the access token is removed
+        response = self.client.post(url)
+        self.assertEqual(response.status_code, 302)
+        self.assertEqual(response.url, '/login/google-oauth2?next=/api/create_project_from_workspace/my-seqr-billing/anvil-no-project-workspace2/submit')
+
+    @responses.activate
+    def test_add_workspace_data(self):
+        # Test insufficient Anvil workspace permission
+        url = reverse(add_workspace_data, args=[PROJECT2_GUID])
+        self.check_collaborator_login(url, login_redirect_url='/login/google-oauth2')
+        response = self.client.post(url, content_type='application/json', data=json.dumps({}))
+        self.assertEqual(response.status_code, 403)
+        self.assertEqual(response.json()['error'], 'Permission Denied')
+
+        # Test requesting to load data from a workspace without an existing project
+        url = reverse(add_workspace_data, args=['no_PROJECT1_GUID'])
+        response = self.client.post(url, content_type='application/json', data=json.dumps({}))
+        self.assertEqual(response.status_code, 404)
+        self.assertEqual(response.json()['error'], 'Project matching query does not exist.')
+
+        # Test missing required fields in the request body
+        url = reverse(add_workspace_data, args=[PROJECT1_GUID])
+        self.client.force_login(self.manager_user)
+
+        self._test_missing_fields_in_request(url, ['uploadedFileId', 'fullDataPath', 'vcfSamples'], TEST_WORKSPACE_NAME)
+
+        self._test_sample_data_error(url)
+
+        self._test_missing_samples(url)
+
+        # Test missing loaded samples
+        self.mock_load_file.return_value = LOAD_SAMPLE_DATA
+        response = self.client.post(url, content_type='application/json', data=json.dumps(REQUEST_BODY))
+        self.assertEqual(response.status_code, 400)
+        self.assertEqual(
+            response.json()['error'],
+            'In order to add new data to this project, new samples must be joint called in a single VCF with all previously'
+            ' loaded samples. The following samples were previously loaded in this project but are missing from the VCF:'
+            ' HG00731, HG00732, HG00733, NA19675_1, NA20870, NA20874')
+
+        # Test a valid operation
+        response = self.client.post(url, content_type='application/json', data=json.dumps(REQUEST_BODY_ADD_DATA))
+        self.assertEqual(response.status_code, 200)
+        response_json = response.json()
+        self.assertEqual(sorted(response_json.keys()), ['familiesByGuid', 'familyNotesByGuid', 'individualsByGuid'])
+        self.assertEqual(sorted(response_json['individualsByGuid'].keys()), ['I0000019_hg00735', 'I000001_na19675', 'I000002_na19678'])
```

Comment:
this is an anti-pattern, for keys we always use `assertSetEqual` instead of sorted lists